### DISCUSSION TITLE
Typescript FieldErrors extended with global form error

### DISCUSSION THREAD

**Discussion by disbelief** [react-hook-form/react-hook-form#4983] [Q&A]
Hi there,

I'm having a really hard time with something that I thought should be straight forward.

Essentially, I have a function that takes an error returned by my server, and I'd like to transpose that into a format that `react-hook-form` can display properly, with `FieldError` objects mapped to specific fields using the [FieldErrors type](https://github.com/react-hook-form/react-hook-form/blob/2dd1e37383a62f8ca4cd13a69d81364d79d1a3fc/src/types/errors.ts#L31).

Some of the errors my server returns are global to the entire form, so I'm trying to add a `base` field to the FieldValues of my form.

Here is my code:

```typescript
interface BaseFormFields {
  base?: any;
};

type WrappedFormFields<TFieldValues> = BaseFormFields & TFieldValues;

function transposeError<WrappedFormFields<T>>(
  customError: CustomError
): FieldErrors<WrappedFormFields<T>> {
  const baseError = customError.base;

  const fieldError: FieldError | undefined = baseError ?? {
    type: 'manual',
    message: baseError,
  };

  return {
    base: fieldError,
  };
}
```

However I'm seeing this error:
```
Type '{ base: FieldError; }' is not assignable to type 'DeepMap<WrappedFormFields<T>, FieldError>'.
```

The `DeepMap<T, TValue>` type [is a bit difficult to parse](https://github.com/react-hook-form/react-hook-form/blob/2dd1e37383a62f8ca4cd13a69d81364d79d1a3fc/src/types/utils.ts#L35), but it seems that as long as the key is present and has a value associated, it should work.

Any help or advice would be really appreciated!

**Comment by bluebill1049** [comment]
we have a feature branch on this one, so you can declare custom error : https://github.com/react-hook-form/react-hook-form/pulls

**Reply by disbelief** [reply]
Thank you, this is precisely what I was looking for. I'd stopped using FieldErrors for my own errors object in the meantime.

**Reply by bluebill1049** [reply]
In fact, for server error, I think it probably makes sense to be tracked separately `useState`?

**Reply by disbelief** [reply]
In some cases yes, that makes sense, eg. a network error or lower-level HTTP error code.

However, my server performs some additional validation on form fields that isn't possible in the client as well. I'd prefer to only pass around a single `errors` object for my form vs two sources of truth about its error state.

**Reply by kotarella1110** [reply]
@disbelief please provide codesandbox 👀
